### ISSUE TITLE
More Node like API

### ISSUE THREAD

**Issue by sindresorhus** [sass/node-sass#676]
The API is currently a bit awkward as it doesn't follow the Node convention.

Instead of:

```js
sass.render({success: function () {}, error: function () {});
```

It should be:

```js
sass.render({}, function (error, result) {});
```

<del>Or embrace promises:

```js
sass.render({}).then(function() {}).catch(function () {});
```
</del>

**Comment by am11** [comment]
We refined API in v2 (https://github.com/sass/node-sass/issues/547), but I agree this is not the end. Promises is the best approach IMO and the syntax will be compatible with PostCSS as well.

Incidentally, with:

```javascript
sass.render({}).then(function() {}).catch(function () {});
```

in `render()` are you proposing to send in `options` object and `then()` will have a success callback?

We would also need to consider custom importers and the upcoming custom functions (https://github.com/sass/node-sass/pull/644).

**Comment by sindresorhus** [comment]
> in render() are you proposing to send in options object and then() will have a success callback?

Yes.

> We would also need to consider custom importers and the upcoming custom functions (#644).

They belong in the options object.

**Comment by am11** [comment]
@sindresorhus, what do you think if we instead go with reactive pattern: http://reactive-extensions.github.io/RxJS/?

**Comment by sindresorhus** [comment]
I like the reactive pattern, but I strongly feel the API should be promise based as it's the standardized describe async operations. Users can easily use any reactive pattern upon a promise interface.

**Comment by dlmanning** [comment]
My two cents: please not promises. Error handling from promises is still kind of fubar. It's trivially easy to wrap the traditional node callback pattern in a promise if your project is using them, but I think it's still premature for a project as widely depended upon as node-sass to adopt them for its api.

**Comment by xzyfer** [comment]
My thoughts here. Leave the existing callback API. I'm aware it's not popular opinion, but it is most accessible.

As @dlmanning points out it's trivial to progressively enhance callbacks to promises, the reverse not so much (it's certainly doable). Arguing that anyone who uses node-sass should be using promises because it's _the right way_ is no argument at all. 

As for the reactive approach, all my points on promises apply, with the addition that it has yet to see wide adoption. Add to that native async/await is coming in ES7, making this the primary API seems premature.

Generally speaking I believe we should aim for providing the lowest developer friction to widest audience, providing sane defaults were possible. If we're going to jump through hoops to support old CentOS I see no reason not to give the attention to the node side of things.

**Comment by sindresorhus** [comment]
I'm fine either either callback or promise.

**Comment by am11** [comment]
I agree with the notion: if it is not broken do not fix it.

But just for sake of argument, what is at risk or what kind of *learning curve* we are anticipating, if node-sass require the `.then()` syntax:

```javascript
sass.render({
  data: 'a{b: c}'
}).then(function(result) {
    // process result
});
```

instead of:

```javascript
sass.render({
  data: 'a{b: c}',
  success: function(result) {
    // process result
  }
});
```

**Comment by sindresorhus** [comment]
Why did you close it? The API is still awkward and not normal Node style callback API...

**Comment by xzyfer** [comment]
I agree with @sindresorhus at the very least the callback signature should follow the node convention of 
```js
function (error, result) {}
```

**Comment by dlmanning** [comment]
node callback convention FTW

**Comment by am11** [comment]
@xzyfer, earlier you were suggesting "Leave the existing callback API" and that there is not point of refactoring..

**Comment by xzyfer** [comment]
Yes. I'm saying leave the callback API. However in the node callback convention is `callback(err, ..rest)`. I was under the impression @sindresorhus was suggesting the argument order was reversed in node-sass (I haven't looked into this myself). 

If that is the case, it should be addressed.

**Comment by kevva** [comment]
node-sass uses an object as callback as opposed to the usual node style `(err, res)`.

```js
sass.render({
  success: function () {},
  error: function () {}
});
```

**Comment by am11** [comment]
@kevva, that looks confusing. What are your thoughts on promise pattern?

**Comment by am11** [comment]
This was addressed by 69a4560 via #684.

**Comment by am11** [comment]
Update to this. We are skipping v2.1 and planning to release v3.0 as the next release instead, so to introduce various breaking changes.

For that matter, we are dropping the old method cb style in favour of node-style cb. 

Credit goes to @Jakobo for https://github.com/sass/node-sass/pull/726.
